### PR TITLE
Optimize cached expressions to avoid redundant guards

### PR DESCRIPTION
**Pull request by smarr**
These changes optimized the dispatch-chain inlining further to avoid redundant guards.

So far, "trivial" expressions have been inlined into the dispatch chain to avoid the overhead of a normal call. However, by simply reusing nodes, there are guards that are redundant with the guards of the dispatch chain.

In this PR, new subclasses for `AbstractDispatchNode` minimize the guards and rely on the check done as part of the dispatch chain.

- field read/write: the new nodes directly access the storage location to avoid redundant guards
- literals: the new node avoids the extra expression node
- object instantiation: the new node avoids the TruffleDSL specialization code, which is not needed with the chain's guards

@OctaveLarose this is probably interesting for you, it avoids a number of redundant guards, but comes with some extra code.

### CODE REVIEW

**Comment by smarr** [comment]
Performance overview: https://rebench.stefan-marr.de/compare/TruffleSOM/fc6d5660735fcab51e70fadf52b33aa3b027b0a3/53d4158dd15306ac0a5e3d90958323e30196304b

SomSom: up to -7% (AST/BC, CE/EE)
NBody: -17% (AST, CE), -18% (AST, EE), -9% (BC, CE), 10% (BC, EE)
Richards: -8% (AST, CE), -9% (AST/BC, EE), -7% (BC, CE)
List: -9% (AST, CE), -11% (AST, EE), -15% (BC, CE/EE)
Towers: -7% (AST, CE/EE), -6% (BC, CE), -10% (BC, EE)

**Review by OctaveLarose — COMMENTED** [review]
Looks good. Are there more guards than `isTrivial()` checks that are avoided through these changes that I don't see, out of curiosity?

**Comment by smarr** [comment]
Ehm, `isTrivial()` is avoided, yes, by refactoring things so that it creates the relevant nodes when possible, and returns `null` otherwise. So, this is more of an optimization to avoid traversing the tree structure more than once.

The relevant guards that are avoided are guards like this: 
- Checking the layout in the dispatch chain and then in the Read*FieldNode again: https://github.com/SOM-st/TruffleSOM/blob/master/src/trufflesom/interpreter/objectstorage/FieldAccessorNode.java#L143
- basically same in the `NewObjectPrim` https://github.com/SOM-st/TruffleSOM/blob/master/src/trufflesom/primitives/basics/NewObjectPrim.java#L22-L23
